### ISSUE TITLE
Mathematic proof of data constancy of cluster members

### ISSUE THREAD

**Issue by RubenKelevra** [ipfs/ipfs-cluster#1004]
Currently there are many Collaborative Cluster ideas popping up, which share one large problem:

Everyone can join. If you don't want to hold every bit of data on every cluster member, you have to trust the members to be trustworthy to hold the data for you.

If you set your minimum redundancy of your pins to 10 a malicious user could add 10 peers with a lot off 'announced' storage. IPFS-Cluster would send every new pin to those 10 clients and they could throw everything away what they receive.

## More effort for cluster joining

-The join to the cluster needs to be more work intense. Every new member needs to calculate a task which cannot be accelerated (much) by using a GPU and would take a regular server several minutes to complete.
-We could ask new members of the cluster to complete such a work task every day hour for the first weeks, to make sure nobody can join a bazillion cluster members with no 'effort'.

This has very low impact on the overall cluster performance, since just new members would be affected. It doesn't use much computing power on the trusted nodes, since they just have to send out the task (with a known outcome) and wait for the results.

The 'trust' value could be stored in the cluster itself, to avoid that a single trusted node outage would lead to lost trust value data.

## Let members proof that they still hold the data

When a trusted node adds data to the cluster it will XOR a many different salts to the stored data, and calculate the checksum for each of them and store it on a local storage.

Those pairs of salt and checksums form a challenge response a node need to be able to answer, when it really holds the data.

The node will get the CID and the salt and need to XOR the salt to the data of the CID and return the checksum back to the trusted node.

A successful answer will raise the trust value for this node.

Since we want to hold redundancies in the cluster, it's necessary to inform all other cluster members beforehand to not provide the CID's data to the node which should be tested. This way the node cannot just pull the data from another node to perform the operation.

If we want to lower the amount of work necessary for the trusted nodes for generating the list of checksums for the challenge response, we could use a deterministic pseudorandom generator to select 1% of the bytes of a CID to XOR and save the seed for the PRNG, the salt (which we XOR) and the corresponding checksum.
  
All operations for the challenge/response are very cheap to do while we have the data already in the memory of the trusted node. But will make it much harder to add a malicious node to a cluster and impact the integrity of the cluster's data.

**Comment by winwisely99** [comment]
I like the trust approach.

It's simple and performant but it still relies on the other nodes also by honouring the semantics of not providing the Cid and Salt.
Also telling the other nodes implies a shared management channel that is s bit like s control plane that is a global shared state. In a way very similar to a global lock.

**Comment by RubenKelevra** [comment]
@winwisely99 thanks!

Right, if we go this route we might increase the latency in certain situations, since the cluster won't provide a given CID for the duration of the challenge/response.

We might want to select just truely random pieces of a pin (depending on the size), to reduce the work for the 'new node' which has to provide the answer. To reduce the latency of the whole operation.

This shouldn't massively affect the probability that a node holds the full amount of data of a pin, since our selection is truely random.

The issue with this approach is, that we need a good source of randomness and also store the random data additionally to the PSNR seed for the 'bitmask', the salt and the checksum.

I think this would increase the overhead too much.

I think it's also a good for the trust value if the challenge response operations generates more I/O. So it's not only a necessary to have a lot of processing power but also a lot of harddrive I/O to emulate nodes.



We should also avoid sending a lot of data to nodes with a low trust value, but enough that they have a chance to earn trust in a reasonable amount of time.

We might want to send them just extra copies of random pins, to fill their space up quickly.

When they earned enough trust, we can drop randomly the surplus of pins from the cluster.

This idea would be possible if this ticket's idea has been added. We would just don't count data on peers with low trust levels:

https://github.com/ipfs/ipfs-cluster/issues/1003

**Comment by hsanjuan** [comment]
> The node will get the CID and the salt and need to XOR the salt to the data of the CID and return the checksum back to the trusted node.
> 
> A successful answer will raise the trust value for this node.
> 

This basically goes down the road of re-implementing Filecoin. The node could do all the calculations and still delete the data and never provide it again. Thus eventually you get to need some sort of Proof of Spacetime and then you are re-doing Filecoin. An approach here is to have cluster peers which are backed by filecoin peers rather than IPFS peers (or both). That, or you simply forget about vetting that every node is doing what it should do and concentrate on building a system in which you are "probably" ok (by doing things more or less randomly).

> Since we want to hold redundancies in the cluster, it's necessary to inform all other cluster members beforehand to not provide the CID's data to the node which should be tested. This way the node cannot just pull the data from another node to perform the operation.

We cannot do this kind of things ("asking everyone for something") with certainity in a distributed system without having full peerset management (i.e. Raft). And that means controlling who joins and who leaves the peerset which is a pain and makes things really difficult for everything else. Plus, malicious nodes could always lie.

The space in which collaborative clusters fit right now is that of putting some piece of data in as many places as possible, which is not a bad thing per-se (having content available in as many places as possible).

IPFS and its pinner component is still missing support for level-N pinning needed to ship IPFS Cluster sharding, but I can imagine that sharding + pinning priority (based on current number of copies of a shard or simply randomness) + politely declining to pin when storage is full would be an approach to get some balance in a system with many people pinning small shards of one big thing.

As a side note, the fact that IPFS will likely move to providing root CIDs only on the DHT is also a problem with the current sharding approach which is agnostic to the DAG layout and in general. This opens the question of whether clusters shall run with their own modified embedded IPFS instance directly.

Also, more or less related to ideas above #524 .

Finally, note that I really like the new ideas proposed and having the associated discussions. Please don't take my party pooping as a signal to stop! It is my task to try steer things towards feasibility both in terms of having something that works and in terms of having something that can actually be attacked with the current, limited, development muscle. These are questions which are not completely new and some thought has already been given to them. Things get very complex, very quickly, sometimes more that can be afforded...

**Comment by RubenKelevra** [comment]
> > The node will get the CID and the salt and need to XOR the salt to the data of the CID and return the checksum back to the trusted node.
> > A successful answer will raise the trust value for this node.
> 
> This basically goes down the road of re-implementing Filecoin. 

> Thus eventually you get to need some sort of Proof of Spacetime and then you are re-doing Filecoin. An approach here is to have cluster peers which are backed by filecoin peers rather than IPFS peers (or both). That, or you simply forget about vetting that every node is doing what it should do and concentrate on building a system in which you are "probably" ok (by doing things more or less randomly).


I haven't really deeply looked at Filecoin, but I like the idea that you can provide storage and bandwidth for getting coins.

But I tend to think of Filecoin as a commercial system. IPFS Collaborative Clusters are more about helping a project out by providing both for free.

Just take a look at all the mirrors for Linux projects, which provides the capability to run updates for your Linux for free.

If those stuff would run on a system like Filecoin we're ending up spending coins for every update we run on our machines.

I don't want to reinvent Filecoin, I just want more reliability for Collaborative Clusters.

> The node could do all the calculations and still delete the data and never provide it again. 

That's partly true.

I don't want to let the node make those calculations right when I store data, but request proofs of random chunks of data from each node from time to time.

Like a ping, just for data availability.

The more responses a node can provide, the more reliable it is and the more we can expect that the node will stick around in the future.

Since the checksums of the data is already readily available, we need to send the node some salt which it has to add to the data before providing an answer.

> > Since we want to hold redundancies in the cluster, it's necessary to inform all other cluster members beforehand to not provide the CID's data to the node which should be tested. This way the node cannot just pull the data from another node to perform the operation.
> 
> We cannot do this kind of things ("asking everyone for something") with certainity in a distributed system without having full peerset management (i.e. Raft). And that means controlling who joins and who leaves the peerset which is a pain and makes things really difficult for everything else. Plus, malicious nodes could always lie.

I'm sure that we can add a lock message to the cluster protocol, send it out to all cluster members except the one we're wanting to challenge.

On the ipfs side we can either implement a connection drop (which would be possible right now), a short blacklist period (which is currently not possible I guess) or add a custom function for the peer commands which add a kind of filter which will drop all requests by the node we want to challenge for the given CID.

This shouldn't be impossible and won't affect the cluster performance much, since we only send out a cid and a node is times cluster members. The cluster members just have to match any requests to a (probably short) list of nodes and cids which should be currently dropped.

Afterwards we send a release lock message to all cluster members.

The only real difference would be on a bootstrap of a node: To avoid that a node of the cluster comes back online and provides instantly the data to the one node we like to challenge, it has to drop all requests of all cluster members until it gets a current version of the lock list from a trusted node.



> The space in which collaborative clusters fit right now is that of putting some piece of data in as many places as possible, which is not a bad thing per-se (having content available in as many places as possible).

Right. I see a potential to improve this, by having just a limited number of copies, to hold large to extremely large amounts of data for scientific purposes for example. Like Boinc is able to provide CPU and GPU time to scientific research, IPFS-Cluster could be the Go-To method of storing the data securely in the internet.

Since we're talking about many dozens of Terabyte the number of people able to provide this amount of storage is pretty limited. So the Collaborative "everyone can join" factor gets pretty small.


> IPFS and its pinner component is still missing support for level-N pinning needed to ship IPFS Cluster sharding, but I can imagine that sharding + pinning priority (based on current number of copies of a shard or simply randomness) + politely declining to pin when storage is full would be an approach to get some balance in a system with many people pinning small shards of one big thing.

I've seen that there's work ongoing doing sharding. But haven't looked in the details.

Question for me: How's that better than just making more than one pin out of the data and reducing the max replication down to like 10 from -1? 🤔

Usually you already got 'files' and 'directories' not 'one massive binary blob of data' you like to store.

> As a side note, the fact that IPFS will likely move to providing root CIDs only on the DHT is also a problem with the current sharding approach which is agnostic to the DAG layout and in general.

So if we would modify the 'ipfs-cluster-ctl add' method to do something like I proposed for the Wikipedia mirror: putting each file / filepath to a x amounts of pins, we should be fine even if IPFS starts just providing root CIDs, right? 🤔

 > This opens the question of whether clusters shall run with their own modified embedded IPFS instance directly.

I dislike the idea of forking IPFS just for running clusters. Everyone should be able to receive stuff which is stored on a cluster without beeing part of the cluster.

Maybe we could just have a flag for pins used by IPFS-Cluster, which will maintain providing 'all' instead of 'root' just for them, until there's a better solution.


> Also, more or less related to ideas above #524 .

I have no idea what this ticket author wanted to convey 🤔



> Finally, note that I really like the new ideas proposed and having the associated discussions. Please don't take my party pooping as a signal to stop! It is my task to try steer things towards feasibility both in terms of having something that works and in terms of having something that can actually be attacked with the current, limited, development muscle. These are questions which are not completely new and some thought has already been given to them. Things get very complex, very quickly, sometimes more that can be afforded...

I really enjoying this community and open communication style in this project, thanks!

**Comment by mohe2015** [comment]
> I'm sure that we can add a lock message to the cluster protocol, send it out to all cluster members except the one we're wanting to challenge.

Unfortunately I don't think proof of storing the data is that simple. For example a malicious user could add 10 nodes to a cluster. Then e.g. two of them should store block X. If node A is challenged node B just tells the cluster it won't provide the data of block X to A but still does so secretly. Therefore you couldn't detect that node A doesn't store block X in any way.


Also if the data gets popular it is probably also stored outside of the cluster and can therefore be simply fetched from there.


> On the ipfs side we can either implement a connection drop (which would be possible right now), a short blacklist period (which is currently not possible I guess) or add a custom function for the peer commands which add a kind of filter which will drop all requests by the node we want to challenge for the given CID.

The problem is that we need to deny access to the block for everyone as the malicious node could otherwise just start a second ipfs-node and use that to download the block from somewhere.

To conclude I don't think it is easy to proof that a node stores some data and I think we should just try to distribute the data to as many nodes as possible. So I think nodes should be able to advertise a specific amount of storage just for that cluster (which is decided by the node operator) and the cluster then allocates as much data as fits in there to that node. Therefore if we have enough benign nodes the malicious ones are not important as they are simply ignored. The current problem just lies in the limit of redundancy as you already said.

P.S. This was just an idea that came off the top of my head...

**Comment by RubenKelevra** [comment]
> > I'm sure that we can add a lock message to the cluster protocol, send it out to all cluster members except the one we're wanting to challenge.
> 
> Unfortunately I don't think proof of storing the data is that simple. For example a malicious user could add 10 nodes to a cluster. Then e.g. two of them should store block X. If node A is challenged node B just tells the cluster it won't provide the data of block X to A but still does so secretly. Therefore you couldn't detect that node A doesn't store block X in any way.

True. That's a possible vector to exploit it, but still easier said than done.

Just adding 10 nodes with 20 EB announced free space to a cluster and dropping every the cluster stores on them is a much easier 'attack'.

This proposal isn't bullet proof, but should raise the bar significantly.

It's like a lock on your door of the house - you can argue that everyone can buy a large hammer in a hardware store and just smash the door.

But we're still don't install 1 inch thick steel doors in our houses.


> Also if the data gets popular it is probably also stored outside of the cluster and can therefore be simply fetched from there.

True but this doesn't really matter, since everything spread around outside will also provide redundancies.

So if a node can fetch the data from outside the cluster and answer the challenge response, I'm fine with that.

But this still requires some amount of work to get this 'attack' working.



> > On the ipfs side we can either implement a connection drop (which would be possible right now), a short blacklist period (which is currently not possible I guess) or add a custom function for the peer commands which add a kind of filter which will drop all requests by the node we want to challenge for the given CID.
> 
> The problem is that we need to deny access to the block for everyone as the malicious node could otherwise just start a second ipfs-node and use that to download the block from somewhere.

Yes, that's right. Don't know how long a challenge response would take to run, but when we're talking about something below 10 seconds, the odds are, that it hardly matters, even when we delay the request.

We could even just drop the connections from which the request is coming in. This would require the client requesting to connect again and hide the delay a bit.

> To conclude I don't think it is easy to proof that a node stores some data and I think we should just try to distribute the data to as many nodes as possible. So I think nodes should be able to advertise a specific amount of storage just for that cluster (which is decided by the node operator) and the cluster then allocates as much data as fits in there to that node. Therefore if we have enough benign nodes the malicious ones are not important as they are simply ignored. The current problem just lies in the limit of redundancy as you already said.
> P.S. This was just an idea that came off the top of my head...

So you're thinking about adding random pins to random nodes until every node has either no storage left or has pinned everything there's to pin? 🤔

When the dataset is large enough, that would result in a static set of pins, since you cannot add anything to the cluster, since everyone has a full space. 

And if you think of the scale of clusters like what Boinc projects are running, for seti@home there are currently around 150k hosts running....


Storing data 150k times if the pinset can fit on all nodes is a bit overkill, wouldn't you agree? 🤔

**Comment by mohe2015** [comment]
> Storing data 150k times if the pinset can fit on all nodes is a bit overkill, wouldn't you agree?

Yeah, I was more thinking about the pacman-cluster. So that if people provide 10GB of storage to it, it will get filled with data. For such a use case that option may be useful as it is simple (KISS-principle) and would work in that case. If the cluster is fast enough less people would probably consider joining it. Therefore the data wouldn't be stored on too many nodes.

> When the dataset is large enough, that would result in a static set of pins, since you cannot add anything to the cluster, since everyone has a full space.

Of course this would need to be thought through more. For example the cluster-master (or how it's called) could tell nodes to unpin some data if there is not enough space left etc. (I don't know how exactly the pinning commands work internally so I may be wrong here). I just imagine that there is some pin priority for every pin and that the cluster-master just publishes that and the nodes decide by themselves randomly which pins they want to hold.

But for larger datasets etc. this is not a good idea.

> > Also if the data gets popular it is probably also stored outside of the cluster and can therefore be simply fetched from there.
>
> True but this doesn't really matter, since everything spread around outside will also provide redundancies.

I agree.


> This proposal isn't bullet proof, but should raise the bar significantly.

> Yes, that's right. Don't know how long a challenge response would take to run, but when we're talking about something below 10 seconds, the odds are, that it hardly matters, even when we delay the request.

I just don't like ideas that are circumventable like the "proof of storage" (I'm really interested in computer security) you proposed especially when they have strange side effects (data not available for a few seconds)


This all seems like a really hard problem and there probably isn't any perfect solution...

**Comment by RubenKelevra** [comment]
> > Storing data 150k times if the pinset can fit on all nodes is a bit overkill, wouldn't you agree?
> 
> Yeah, I was more thinking about the pacman-cluster. So that if people provide 10GB of storage to it, it will get filled with data. For such a use case that option may be useful as it is simple (KISS-principle) and would work in that case. If the cluster is fast enough less people would probably consider joining it. Therefore the data wouldn't be stored on too many nodes.

Yeah that's actually what's happening right now, but we don't have a limit per collab cluster, but just a total limit.

I've set a minimum limit of 1, which means I can add data to the cluster, regardless if the server sees other peers or not.

Maximum limit is 10, so when there are more than 10 nodes in the cluster, only 10 would taught to pin the the newly pinned data.

Cluster nodes with the most free space will be selected first.

So if a person looks at the config and sees the maximum is 10, he can join 10 clients with massive amounts of free space and would get all data allocated - without any check it he reliability would return data he stored on his nodes.

If he leaves the cluster with all nodes simultaneously, all data he hold is gone.

This only works at the time I send the pin/add command. So if you join your node right now, you won't receive old data from the cluster.

When I set -1 as replication, the cluster would send you the whole ~60 GB until your storage is full - which would be quite random.

If there's stuff unpinned from the cluster, the garbage collection could remove and reclaim the storage from your node and you would randomly fetch new stuff. Until the storage is full again.


If I set the minimum to 2 or more, the cluster will watch if a node goes offline and the redundancies of the pins falls below $minimum and would initiate a repinning event for all pins which are now below the minimum, which is somewhat expensive to run.

If the node returns it will just provide an additional copy and clean up the state again, so basically remove everything which is unpinned by the cluster.

For the Arch Linux cluster it isn't really important how many people join the cluster, since everyone using the pkg cache will also cache the packages for the whole network.

So packages like the kernel are likely to be cached on every node using this system.

Additionally we don't care if stuff in the cluster is gone, I hold a full copy of the data on the server anyway - outside from IPFS to run rsync against it for updates.

The script which syncs both storages is resilient against sudden disappearing of files from IPFS, so it would just add the new version and don't care if the old one is gone.

The db files are updated extremely often, so it will lay in the cache of the server at all time, and if it's really gone, it just take some hours max to have it replaced by a new version.

When you use the scripts I wrote (which are not really released yet), they will just copy a new db file from ipfs when they are available. If not it just cancels the sync.

You can still update packages using the last db file you synched and if ipfs doesn't hold a file, pacman will default back to pulling it from a regular mirror.

When files are pulled from a regular mirror, the next sync operation will add them to the local ipfs cache (without pinning) to make use of the already spend bandwidth.

Ipfs will automatically remove stuff from the cache when necessary, because the storage is full. 


> > When the dataset is large enough, that would result in a static set of pins, since you cannot add anything to the cluster, since everyone has a full space.
> 
> Of course this would need to be thought through more. For example the cluster-master (or how it's called) could tell nodes to unpin some data if there is not enough space left etc. (I don't know how exactly the pinning commands work internally so I may be wrong here). I just imagine that there is some pin priority for every pin and that the cluster-master just publishes that and the nodes decide by themselves randomly which pins they want to hold.
>
> But for larger datasets etc. this is not a good idea.

There's no 'cluster-master', just one or more nodes which has write access to the cluster - trusted nodes. The cluster itself has just peers, there's nothing different between them. They can work together without a trusted node online - they just cannot change the content.

Pinning from a trusted node is pretty simple: They can release an update of a data file which is basically a list of pins and where it should be placed.

While running the command the trusted node looks at the metric of nodes and selects as many nodes as $maxredundancy is set to, beginning from the nodes which has the most space left.

When it reaches the end of the list, and $minimumredundancy isn't met yet, the pinning operations fails.

If it reaches the limit and the $maximumredundancy isn't met, it will be ignored and the pin will be stored on less than the set maximum redundancy.

When you run a pin operation, the data is already somewhere in the network available, so the nodes of the cluster will receive the new version of the data file, which holds which pin should be stored where, and if they are mentioned, they will a local pinning operation. So IPFS will just fetch the data from the network and store it in the cache.

If you run a add operation, you just add the data basically to your local client's cache first (when you use --local) and the other nodes will pull the data from your cache.

You can also ommit the --local flag, in this case the data is send to the remote nodes which should add it. (Have no idea how this works in detail).



> > This proposal isn't bullet proof, but should raise the bar significantly.
> 
> > Yes, that's right. Don't know how long a challenge response would take to run, but when we're talking about something below 10 seconds, the odds are, that it hardly matters, even when we delay the request.
> 
> I just don't like ideas that are circumventable like the "proof of storage" (I'm really interested in computer security) you proposed especially when they have strange side effects (data not available for a few seconds)
> 
> This all seems like a really hard problem and there probably isn't any perfect solution...

I don't think there's a solution to this, which covers all bases, as long as we want cluster peers to exchange the data.

We could solve this by sending each cluster peer different data, which he doesn't know how it's mixed up, like with a redundancy adding codec plus some kind of scrambling or encryption.

But this would result in the matter, that nobody except the trusted nodes could decode the data again.

I might be completely wrong here - there's some amazing math out there. 😙🎶

---

Keep everything I wrote with a grain of salt, I'm new to this project and just read the docs and experimented a bit with this software. There might be major errors in my statements.

**Comment by hsanjuan** [comment]
> I don't want to let the node make those calculations right when I store data, but request proofs of random chunks of data from each node from time to time.

Essentially Filecoin.

> I'm sure that we can add a lock message to the cluster protocol, send it out to all cluster members except the one we're wanting to challenge.

Still relies on cluster members behaving well, on network connectivity, on low latencies and on knowledge of the full peerset at a certain point in time.

> I see a potential to improve this, by having just a limited number of copies, to hold large to extremely large amounts of data for scientific purposes for example. Like Boinc is able to provide CPU and GPU time to scientific research, IPFS-Cluster could be the Go-To method of storing the data securely in the internet.

I see the potential, I just don't see this approach as a good way to do it.

> Question for me: How's that better than just making more than one pin out of the data and reducing the max replication down to like 10 from -1?

It is better in the sense that you don't have to figure out how to split your data and the cluster does it for you in a data-agnostic way.

>  we should be fine even if IPFS starts just providing root CIDs, right?

Probably, but you cannot really recursively pin to original root at all. Every pin will store everything hanging down from it, so you have to literally split your DAG in disjoint parts and make sure you are not pulling a full other branch that you did not want to pin in the recursive traversal. The approach in the end is very specific to the data being handled.

> I dislike the idea of forking IPFS just for running clusters. Everyone should be able to receive stuff which is stored on a cluster without beeing part of the cluster.

It's not forking IPFS, it is running ipfs with an optimized configuration/components. It would still be in the normal network and interact with the standalone peers.

> Maybe we could just have a flag for pins used by IPFS-Cluster, which will maintain providing 'all' instead of 'root' just for them, until there's a better solution.

If you run the math, it is really unsustainable to provide every single block for big things. The stress it puts on the nodes and on the DHT in general is not good.

>>     Also, more or less related to ideas above #524 .
> I have no idea what this ticket author wanted to convey thinking

Sorry: It's #966 .

> > To conclude I don't think it is easy to proof that a node stores some data and I think we should just try to distribute the data to as many nodes as possible. So I think nodes should be able to advertise a specific amount of storage just for that cluster (which is decided by the node operator) and the cluster then allocates as much data as fits in there to that node. Therefore if we have enough benign nodes the malicious ones are not important as they are simply ignored. The current problem just lies in the limit of redundancy as you already said.
> > P.S. This was just an idea that came off the top of my head...
> 
> So you're thinking about adding random pins to random nodes until every node has either no storage left or has pinned everything there's to pin? thinking
> 
> When the dataset is large enough, that would result in a static set of pins, since you cannot add anything to the cluster, since everyone has a full space.
> 
> And if you think of the scale of clusters like what Boinc projects are running, for seti@home there are currently around 150k hosts running....
> 
> Storing data 150k times if the pinset can fit on all nodes is a bit overkill, wouldn't you agree?

I agree with `I think nodes should be able to advertise a specific amount of storage just for that cluster (which is decided by the node operator) and the cluster then allocates as much data as fits in there to that node`. 

And yes, nodes would have a list of things to pin. Now they have to decide (themselves, without any orders or coordination from others), what items in that list they actually store based on some priority metric that they put together.

i.e. You can optimize to store as much as possible, or you can make DHT-queries and try to store the most under-provided items etc. But that logic is in those peers only and needs no coordination with the rest. Also, for unixfs the DAG-nodes carry information about how big the underlying DAG is, which allows for sharding dag-branches into pieces small enough etc, but that is only for UnixFS, not for other type of DAGs or custom formats.
